File: software/examples/c/example_callback.c

#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_barometer.h"

#define HOST "localhost"
#define PORT 4223
#define UID "bAc" // Change to your UID

// Callback function for air pressure callback (parameter has unit mbar/1000)
void cb_air_pressure(int32_t air_pressure) {
	printf("Air Pressure: %f mbar\n", air_pressure/1000.0);
}

// Callback function for altitude callback (parameter has unit cm)
void cb_altitude(int32_t altitude) {
	printf("Altitude: %f m\n", altitude/100.0);
}

int main() {
	// Create IP connection to brickd
	IPConnection ipcon;
	if(ipcon_create(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not create connection\n");
		exit(1);
	}

	// Create device object
	Barometer b;
	barometer_create(&b, UID);

	// Add device to IP connection
	if(ipcon_add_device(&ipcon, &b) < 0) {
		fprintf(stderr, "Could not connect to Bricklet\n");
		exit(1);
	}
	// Don't use device before it is added to a connection

	// Set Period for air pressure and altitude callbacks to 1s (1000ms)
	// Note: The air pressure and altitude callbacks are only called every second
	//       if the air pressure or altitude has changed since the last call!
	barometer_set_air_pressure_callback_period(&b, 1000);
	barometer_set_altitude_callback_period(&b, 1000);

	// Register air pressure callback to function cb_air_pressure
	barometer_register_callback(&b,
	                            BAROMETER_CALLBACK_AIR_PRESSURE,
	                            cb_air_pressure);

	// Register altitude callback to function cb_altitude
	barometer_register_callback(&b,
	                            BAROMETER_CALLBACK_ALTITUDE,
	                            cb_altitude);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon);
}

File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{barometer_bricklet::*, ipconnection::IpConnection};

const HOST: &str = "127.0.0.1";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Barometer Bricklet

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection
    let barometer_bricklet = BarometerBricklet::new(UID, &ipcon); // Create device object

    ipcon.connect(HOST, PORT).recv()??; // Connect to brickd
                                        // Don't use device before ipcon is connected

    //Create listener for air pressure events.
    let air_pressure_listener = barometer_bricklet.get_air_pressure_receiver();
    // Spawn thread to handle received events. This thread ends when the barometer_bricklet
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for event in air_pressure_listener {
            println!("Air Pressure: {}{}", event as f32 / 1000.0, " mbar");
        }
    });

    // Set period for air pressure listener to 1s (1000ms)
    // Note: The air pressure callback is only called every second
    //       if the air pressure has changed since the last call!
    barometer_bricklet.set_air_pressure_callback_period(1000);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}
